File: ios/GrowOp/Services/ServerConfig.swift
import Foundation

/// How the app reaches the grow brain.
enum ConnectionMode: String, CaseIterable, Codable {
    /// Plain HTTP to the add-on's own port on the home LAN.
    case direct
    /// Through Home Assistant's add-on ingress proxy (works remotely via Nabu Casa).
    case homeAssistant

    var title: String {
        switch self {
        case .direct: return "Same Wi‑Fi"
        case .homeAssistant: return "Through Home Assistant"
        }
    }
}

/// Where the grow brain lives. Stored in UserDefaults (single-user app).
struct ServerConfig: Equatable {
    var mode: ConnectionMode = .direct
    /// Direct mode: the add-on's base URL, e.g. http://homeassistant.local:8099
    var baseURL: String = ""
    /// Grow Brain API key (sent as X-API-Key in BOTH modes).
    var apiKey: String = ""
    /// Home Assistant mode: HA base URL, e.g. https://xxxx.ui.nabu.casa
    var haURL: String = ""
    /// Home Assistant long-lived access token.
    var haToken: String = ""
    /// Cached discovery results (refreshed automatically when stale).
    var haAddonSlug: String? = nil
    var haIngressPath: String? = nil

    static let defaultURL = "http://homeassistant.local:8099"

    private enum Keys {
        static let mode = "growop.server.mode"
        static let url = "growop.server.url"
        static let apiKey = "growop.server.apiKey"
        static let haURL = "growop.server.haURL"
        static let haToken = "growop.server.haToken"
        static let haAddonSlug = "growop.server.haAddonSlug"
        static let haIngressPath = "growop.server.haIngressPath"
    }

    static func load() -> ServerConfig {
        let d = UserDefaults.standard
        // Provisioning: values passed as launch arguments (`-growop.server.mode homeAssistant -growop.server.haURL …`)
        // live only in the volatile argument domain; persist them once so the app stays configured after relaunch.
        let args = d.volatileDomain(forName: UserDefaults.argumentDomain)
        if args[Keys.apiKey] != nil, (args[Keys.url] != nil || args[Keys.haURL] != nil) {
            let provisioned = ServerConfig(
                mode: ConnectionMode(rawValue: args[Keys.mode] as? String ?? "") ?? .direct,
                baseURL: args[Keys.url] as? String ?? "",
                apiKey: args[Keys.apiKey] as? String ?? "",
                haURL: args[Keys.haURL] as? String ?? "",
                haToken: args[Keys.haToken] as? String ?? "",
                haAddonSlug: nil, haIngressPath: nil)
            provisioned.save()
            d.removeVolatileDomain(forName: UserDefaults.argumentDomain)
            d.setVolatileDomain([:], forName: UserDefaults.argumentDomain)
            return provisioned
        }
        return ServerConfig(
            mode: ConnectionMode(rawValue: d.string(forKey: Keys.mode) ?? "") ?? .direct,
            baseURL: d.string(forKey: Keys.url) ?? "",
            apiKey: d.string(forKey: Keys.apiKey) ?? "",
            haURL: d.string(forKey: Keys.haURL) ?? "",
            haToken: d.string(forKey: Keys.haToken) ?? "",
            haAddonSlug: d.string(forKey: Keys.haAddonSlug),
            haIngressPath: d.string(forKey: Keys.haIngressPath)
        )
    }

    func save() {
        let d = UserDefaults.standard
        d.set(mode.rawValue, forKey: Keys.mode)
        d.set(baseURL, forKey: Keys.url)
        d.set(apiKey, forKey: Keys.apiKey)
        d.set(haURL, forKey: Keys.haURL)
        d.set(haToken, forKey: Keys.haToken)
        d.set(haAddonSlug, forKey: Keys.haAddonSlug)
        d.set(haIngressPath, forKey: Keys.haIngressPath)
    }

    static func clear() {
        let d = UserDefaults.standard
        [Keys.mode, Keys.url, Keys.apiKey, Keys.haURL, Keys.haToken, Keys.haAddonSlug, Keys.haIngressPath]
            .forEach { d.removeObject(forKey: $0) }
    }

    /// Adds a scheme if missing (http:// for direct, https:// for HA) and strips trailing slashes.
    static func normalize(_ raw: String, defaultScheme: String) -> String {
        var s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if !s.isEmpty, !s.lowercased().hasPrefix("http://"), !s.lowercased().hasPrefix("https://") {
            s = defaultScheme + "://" + s
        }
        while s.hasSuffix("/") { s.removeLast() }
        return s
    }

    /// Direct-mode base URL, normalised.
    var normalizedBaseURL: String { Self.normalize(baseURL, defaultScheme: "http") }

    /// Home Assistant base URL, normalised.
    var normalizedHAURL: String { Self.normalize(haURL, defaultScheme: "https") }

    var trimmedAPIKey: String { apiKey.trimmingCharacters(in: .whitespacesAndNewlines) }
    var trimmedHAToken: String { haToken.trimmingCharacters(in: .whitespacesAndNewlines) }

    var isConfigured: Bool {
        guard !trimmedAPIKey.isEmpty else { return false }
        switch mode {
        case .direct: return !normalizedBaseURL.isEmpty
        case .homeAssistant: return !normalizedHAURL.isEmpty && !trimmedHAToken.isEmpty
        }
    }

    /// True when the user-entered fields differ (ignores cached discovery values).
    func differsInUserFields(from other: ServerConfig) -> Bool {
        mode != other.mode || baseURL != other.baseURL || apiKey != other.apiKey
            || haURL != other.haURL || haToken != other.haToken
    }
}

File: android/app/src/main/java/com/growop/app/data/ServerConfig.kt
package com.growop.app.data

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first

/** How the app reaches the grow brain. */
enum class ConnectionMode(val title: String) {
    /** Plain HTTP to the add-on's own port on the home LAN. */
    DIRECT("Same Wi‑Fi"),
    /** Through Home Assistant's add-on ingress proxy (works remotely via Nabu Casa). */
    HOME_ASSISTANT("Through Home Assistant");

    companion object {
        fun from(s: String?): ConnectionMode = entries.firstOrNull { it.name == s } ?: DIRECT
    }
}

/** Where the grow brain lives. */
data class ServerConfig(
    val mode: ConnectionMode = ConnectionMode.DIRECT,
    /** Direct mode: the add-on's base URL, e.g. http://homeassistant.local:8099 */
    val baseUrl: String = "",
    /** Grow Brain API key (sent as X-API-Key in BOTH modes). */
    val apiKey: String = "",
    /** Home Assistant mode: HA base URL, e.g. https://xxxx.ui.nabu.casa */
    val haUrl: String = "",
    /** Home Assistant long-lived access token. */
    val haToken: String = "",
    /** Cached discovery results (refreshed automatically when stale). */
    val haAddonSlug: String? = null,
    val haIngressPath: String? = null,
) {
    companion object {
        const val DEFAULT_URL = "http://homeassistant.local:8099"

        /** Adds a scheme if missing and strips trailing slashes. */
        fun normalize(raw: String, defaultScheme: String): String {
            var s = raw.trim()
            if (s.isNotEmpty() && !s.lowercase().startsWith("http://") && !s.lowercase().startsWith("https://")) {
                s = "$defaultScheme://$s"
            }
            return s.trimEnd('/')
        }
    }

    val normalizedBaseUrl: String get() = normalize(baseUrl, "http")
    val normalizedHaUrl: String get() = normalize(haUrl, "https")
    val trimmedApiKey: String get() = apiKey.trim()
    val trimmedHaToken: String get() = haToken.trim()

    val isConfigured: Boolean
        get() {
            if (trimmedApiKey.isEmpty()) return false
            return when (mode) {
                ConnectionMode.DIRECT -> normalizedBaseUrl.isNotEmpty()
                ConnectionMode.HOME_ASSISTANT -> normalizedHaUrl.isNotEmpty() && trimmedHaToken.isNotEmpty()
            }
        }

    /** True when the user-entered fields differ (ignores cached discovery values). */
    fun differsInUserFields(other: ServerConfig): Boolean =
        mode != other.mode || baseUrl != other.baseUrl || apiKey != other.apiKey ||
            haUrl != other.haUrl || haToken != other.haToken
}

private val Context.growDataStore: DataStore<Preferences> by preferencesDataStore(name = "growop")

/** DataStore-backed persistence for the connection config and small per-phone preferences. */
class ConfigStore(private val context: Context) {
    private object Keys {
        val mode = stringPreferencesKey("server.mode")
        val url = stringPreferencesKey("server.url")
        val apiKey = stringPreferencesKey("server.apiKey")
        val haUrl = stringPreferencesKey("server.haUrl")
        val haToken = stringPreferencesKey("server.haToken")
        val haAddonSlug = stringPreferencesKey("server.haAddonSlug")
        val haIngressPath = stringPreferencesKey("server.haIngressPath")
        val myPlantId = intPreferencesKey("myPlantId")
        val units = stringPreferencesKey("units")
    }

    data class Loaded(val config: ServerConfig, val myPlantId: Int?, val units: String?)

    suspend fun load(): Loaded {
        val p = context.growDataStore.data.first()
        val cfg = ServerConfig(
            mode = ConnectionMode.from(p[Keys.mode]),
            baseUrl = p[Keys.url] ?: "",
            apiKey = p[Keys.apiKey] ?: "",
            haUrl = p[Keys.haUrl] ?: "",
            haToken = p[Keys.haToken] ?: "",
            haAddonSlug = p[Keys.haAddonSlug],
            haIngressPath = p[Keys.haIngressPath],
        )
        return Loaded(cfg, p[Keys.myPlantId], p[Keys.units])
    }

    suspend fun saveConfig(cfg: ServerConfig) {
        context.growDataStore.edit { p ->
            p[Keys.mode] = cfg.mode.name
            p[Keys.url] = cfg.baseUrl
            p[Keys.apiKey] = cfg.apiKey
            p[Keys.haUrl] = cfg.haUrl
            p[Keys.haToken] = cfg.haToken
            cfg.haAddonSlug?.let { p[Keys.haAddonSlug] = it } ?: p.remove(Keys.haAddonSlug)
            cfg.haIngressPath?.let { p[Keys.haIngressPath] = it } ?: p.remove(Keys.haIngressPath)
        }
    }

    suspend fun clearConfig() {
        context.growDataStore.edit { p ->
            listOf(Keys.mode, Keys.url, Keys.apiKey, Keys.haUrl, Keys.haToken, Keys.haAddonSlug, Keys.haIngressPath)
                .forEach { p.remove(it) }
        }
    }

    suspend fun saveMyPlantId(id: Int?) {
        context.growDataStore.edit { p -> if (id != null) p[Keys.myPlantId] = id else p.remove(Keys.myPlantId) }
    }

    suspend fun saveUnits(units: String) {
        context.growDataStore.edit { p -> p[Keys.units] = units }
    }
}
